File: latihan/2-inheritance/Bus.cpp
/*  NIM     : 13521129
    Nama    : Chiquita Ahsanunnisa
    Tanggal : 19 Februari 2023
    Topik   : Latihan Inheritance - Bus.cpp
*/

#include "Bus.hpp"

// BUS (DERIVED CLASS)

// CTOR
// Bus::Bus() : Bus(0, 0, "XXX", 0) {} // ERROR C++11
Bus::Bus() : Kendaraan(0,0,"XXX"), kapasitas(0) {}
Bus::Bus(int n, int t, string m, int k) : Kendaraan(n,t,m), kapasitas(k) {}

// CCTOR tidak butuh karena tidak ada DMA
// Bus::Bus(const Bus& b) {}

// ASSIGNMENT tidak butuh karena tidak ada DMA
// Bus& Bus::operator=(const Bus& b) {}

// DTOR tidak butuh karena tidak ada DMA baik di base maupun derived class
// Bus::~Bus() {}

// redefine virtual method
void Bus::printInfo() const {
    Kendaraan::printInfo();
    cout << "Kategori: Bus" << endl;
}
int Bus::biayaSewa(int lamaSewa) const {
    return 1000000 * lamaSewa;
}

// VIRTUAL COPY
Kendaraan* Bus::clone() const {
    return new Bus(*this);
}

File: latihan/5-java-inheritance/Bus.java
/*  NIM     : 13521129
    Nama    : Chiquita Ahsanunnisa
    Tanggal : 31 Maret 2023
    Topik   : Latihan Java Inheritance - Bus.java
*/

public class Bus extends Kendaraan {
    private int kapasitas;

    // CTOR
    public Bus() {
        super();
        kapasitas = 0;
    }
    public Bus(int nomor, int tahun, String merk, int kapasitas) {
        super(nomor, tahun, merk);
        this.kapasitas = kapasitas;
    }

    // OTHERS
    @Override
    public void printInfo() {
        super.printInfo();
        System.out.println("Kategori       : Bus");
        System.out.println("Kapasitas      : " + kapasitas);
    }
    @Override
    public int biayaSewa(int lamaSewa) { return 1000000 * lamaSewa; }
}
